src/Sort.c

```diff
@@ -144,6 +156,182 @@ int quickSort(int arr[], int low, int high){
+      break;
+    } else if (charset[i] == charset2[i]) {
+      result = 0;
+      // continue looking for the bigger letter
+    } else {
+      result = 1;
+      // break here if first letter was smaller
+      break;
+    }
+  }
+  return result;
+}
+
+int partitionWithElementsByLetters(Element arr[], int low, int high) {
+  // last Element becomes pivot
+  char *pivot = (arr+high)->letters;
+  // i is temporary pivot index - marks the index in the array, where everything with a smaller index (to the left)
+  // is already smaller than the pivot itself, but not sorted itself
+  int i = (low - 1);
+
+  // for all elements smaller than pivot move them to the left
+  for (int j = low; j < high; j++){
+    if (isBiggerOrSameCharset(pivot, (arr+j)->letters) == 0){
+       // move temporary pivot index forward
+      i++;
+      // swap current element with the element at the temporary pivot
+      swapElements(arr+i, arr+j);
+    }
+  }
+  swapElements(arr+i+1, arr+high);
+  return (i+1);
+}
+
+// executes quick sort on array of Element
+// either sorting it by numbers or letters depending on kind (either 'l' or 'n')
+int quickSortWithElements(Element arr[], int low, int high, char *kind){
+  if (low < high) {
+    int pivot;
+    // TODO is there a better way to differentiate between those two ways of sorting?
+    if (strcmp(kind, "n") == 0) {
+      // partition array and get pivot index sorted by numbers
+      pivot = partitionWithElementsByNumber(arr, low, high);
+    } else if (strcmp(kind, "l") == 0) {
+      // partition array and get pivot index sorted by letters
+      pivot = partitionWithElementsByLetters(arr, low, high);
+    }
+    // Sort the two partitions
+    quickSortWithElements(arr, low, pivot - 1, kind); // sort left side of pivot
+    quickSortWithElements(arr, pivot + 1, high, kind); // sort right side of pivot
+  }
+  return 0;
+}
+
+// searches for given integer in and array of Element of given size already sorted by numbers
+// starts searching in the middle of array and goes left or right depending on target element
+int searchForNumber(Element arr[], int target, int size) {
+  // first choose middle of array
+  int middle;
+  // if size is even, middle is offset by 1 to the right
+  if (size % 2 == 0) {
+    middle = size / 2;
+  } else {
+    middle = (size - 1) / 2;
+  }
+  if (size > 0) {
+    // check if selected middle is already the target element
+    if ((arr+middle)->number == target) {
+      printf("Found it!\n");
+      printf("Number: %d - Letters: %s\n", (arr+middle)->number, (arr+middle)->letters);
+    } // if target is bigger than middle call searchForNumber on right half of the array
+    else if ((arr+middle)->number < target) {
+      searchForNumber(arr+middle, target, size - middle);
+    } // if target is smaller than middle call searchForNumber on left half of the array
+    else if ((arr+middle)->number > target) {
+      searchForNumber(arr, target, middle);
+    }
+  }
+  return 0;
```

Comment:
since this is a recursive function, it might be clearer if the exit condition is at the very beginning, i.e.
if (size <= 0) {
return 0}
then the rest of what the function call does.
include print statement if the Element was not found.